### PR TITLE
HTTP Validator API:  `/eth/v1/keystores`

### PR DESCRIPTION
**Pull request by james-prysm**
**What type of PR is this?**

Other

**What does this PR do? Why is it needed?**

This PR implements the pure http REST `representation of /eth/v1/keystores` with `GET`,`POST` and `DELETE`

- also fully removes the API middleware from the validator client
- adding in CORs options for REST endpoints is still a todo
- keymanager APIs will not require the bearer token anymore

example results
```
curl -X GET "http://localhost:7500/eth/v1/keystores" -H "accept: application/json"
{"data":[{"validating_pubkey":"0xaa0ef7404c3a164741856c78cc844072448886291e3c5b29514001d1bd8c5ccc9c342a1caa58c3b45734195ef4b4122c","derivation_path":""},{"validating_pubkey":"0x8b91f34c039c348f73c2dda492998e268c2815a433f59a6667258267739947dcfc8ade6823b3f1f3cfef824497eb113c","derivation_path":""}]}

curl -X DELETE "http://localhost:7500/eth/v1/keystores" -H  "accept: application/json" -H  "Content-Type: application/json" -d "{\"pubkeys\":[\"0x8b91f34c039c348f73c2dda492998e268c2815a433f59a6667258267739947dcfc8ade6823b3f1f3cfef824497eb113c\"]}"

{"data":[{"status":"DELETED","message":""}],"slashing_protection":"{\"metadata\":{\"interchange_format_version\":\"5\",\"genesis_validators_root\":\"0x043db0d9a83813551ee2f33450d23797757d430911a9320530ad8a0eabc43efb\"},\"data\":[{\"pubkey\":\"0x8b91f34c039c348f73c2dda492998e268c2815a433f59a6667258267739947dcfc8ade6823b3f1f3cfef824497eb113c\",\"signed_blocks\":[],\"signed_attestations\":[]}]}"}

curl -X GET "http://localhost:7500/eth/v1/keystores" -H "accept: application/json"
{"data":[{"validating_pubkey":"0xaa0ef7404c3a164741856c78cc844072448886291e3c5b29514001d1bd8c5ccc9c342a1caa58c3b45734195ef4b4122c","derivation_path":""}]}

curl -X POST 'http://localhost:7500/eth/v1/keystores' \
-H 'Content-Type: application/json' \
-d '{
  "keystores": [
    "{\"crypto\": {\"kdf\": {\"function\": \"scrypt\", \"params\": {\"dklen\": 32, \"n\": 262144, \"r\": 8, \"p\": 1, \"salt\": \"5e1b03cc37210793c1e5db68fe7ab39c914b17659e62f24739f6c206584da9c8\"}, \"message\": \"\"}, \"checksum\": {\"function\": \"sha256\", \"params\": {}, \"message\": \"1b8425cd92ac700cd363ea1862979c8886ece1ae0ddd8be1602a01752071fed7\"}, \"cipher\": {\"function\": \"aes-128-ctr\", \"params\": {\"iv\": \"1f34217a64cc3e9635b59fe5296ed9ac\"}, \"message\": \"667c7b4c01f7e96b4f29a233bc234561XXX66511f979e966f25bd685a66cff659\"}}, \"description\": \"\", \"pubkey\": \"8b91f34c039c348f73c2dda492998e268c2815a433f59a6667258267739947dcfc8ade6823b3f1f3cfef824497eb113c\", \"path\": \"m/12381/3600/2/0/0\", \"uuid\": \"5a740f4c-cb1c-41f8-8340-a8cbf890582d\", \"version\": 4}"
  ],
  "passwords":["placeholderXXX"]
}'

{"data":[{"status":"IMPORTED","message":""}]}

curl -X GET "http://localhost:7500/eth/v1/keystores" -H "accept: application/json"
{"data":[{"validating_pubkey":"0xaa0ef7404c3a164741856c78cc844072448886291e3c5b29514001d1bd8c5ccc9c342a1caa58c3b45734195ef4b4122c","derivation_path":""},{"validating_pubkey":"0x8b91f34c039c348f73c2dda492998e268c2815a433f59a6667258267739947dcfc8ade6823b3f1f3cfef824497eb113c","derivation_path":""}]}
```

**Which issues(s) does this PR fix?**

Part of #13041

### CODE REVIEW

**Comment by james-prysm** [comment]
something happened there's an issue while retesting

**Comment by james-prysm** [comment]
ready again

File: validator/accounts/accounts_delete.go [hunk 1]
```diff
@@ -26,7 +25,7 @@ func (acm *CLIManager) Delete(ctx context.Context) error {
 		if len(acm.filteredPubKeys) == 1 {
 			promptText := "Are you sure you want to delete 1 account? (%s) Y/N"
 			resp, err := prompt.ValidatePrompt(
-				os.Stdin, fmt.Sprintf(promptText, au.BrightGreen(formattedPubKeys[0])), prompt.ValidateYesOrNo,
+				acm.inputReader, fmt.Sprintf(promptText, au.BrightGreen(formattedPubKeys[0])), prompt.ValidateYesOrNo,
```
Comment (rkapka) [inline validator/accounts/accounts_delete.go:28]: Is this change related?

File: validator/accounts/accounts_delete.go [hunk 2]
```diff
@@ -41,7 +40,7 @@ func (acm *CLIManager) Delete(ctx context.Context) error {
 			} else {
 				promptText = fmt.Sprintf(promptText, len(acm.filteredPubKeys), au.BrightGreen(allAccountStr))
 			}
-			resp, err := prompt.ValidatePrompt(os.Stdin, promptText, prompt.ValidateYesOrNo)
+			resp, err := prompt.ValidatePrompt(acm.inputReader, promptText, prompt.ValidateYesOrNo)
```
Comment (rkapka) [inline validator/accounts/accounts_delete.go:43]: Same here

File: validator/node/node.go [hunk 3]
```diff
@@ -821,15 +818,13 @@ func (c *ValidatorClient) registerRPCGatewayService(router *mux.Router) error {
 		),
 		gwruntime.WithForwardResponseOption(gateway.HttpResponseModifier),
 	)
-	muxHandler := func(apiMware *apimiddleware.ApiProxyMiddleware, h http.HandlerFunc, w http.ResponseWriter, req *http.Request) {
+
+	// TODO: remove api middleware from this
```
Comment (rkapka) [inline validator/node/node.go:822]: I would remove this TODO because the code will fail to compile once we remove the middleware parameter from `WithMuxHandler`

File: validator/node/node.go [hunk 4]
```diff
@@ -821,15 +818,13 @@ func (c *ValidatorClient) registerRPCGatewayService(router *mux.Router) error {
...
+	muxHandler := func(_ *apimiddleware.ApiProxyMiddleware, h http.HandlerFunc, w http.ResponseWriter, req *http.Request) {
 		// The validator gateway handler requires this special logic as it serves two kinds of APIs, namely
-		// the standard validator keymanager API under the /eth namespace, and the Prysm internal
 		// validator API under the /api namespace. Finally, it also serves requests to host the validator web UI.
```
Comment (rkapka) [inline validator/node/node.go:825]: Are these comments still correct?

File: validator/node/node.go [hunk 5]
```diff
@@ -821,15 +818,13 @@ func (c *ValidatorClient) registerRPCGatewayService(router *mux.Router) error {
...
-		if strings.HasPrefix(req.URL.Path, "/api/eth/") {
-			req.URL.Path = strings.Replace(req.URL.Path, "/api", "", 1)
-			// If the prefix has /eth/, we handle it with the standard API gateway middleware.
-			apiMware.ServeHTTP(w, req)
-		} else if strings.HasPrefix(req.URL.Path, "/api") {
+		// note: keymanager api endpoints are handled by reregistering all the route handlers before the catchall
```
Comment (rkapka) [inline validator/node/node.go:826]: I don't understand this. Can you explain this in more detail?

File: validator/rpc/handlers_keymanager.go [hunk 6]
```diff
@@ -19,11 +21,290 @@ import (
...
+	}
+	keystores := make([]*keymanager.Keystore, len(req.Keystores))
+	for i := 0; i < len(req.Keystores); i++ {
+		k := &keymanager.Keystore{}
+		err = json.Unmarshal([]byte(req.Keystores[i]), k)
+		if k.Description == "" && k.Name != "" {
+			k.Description = k.Name
+		}
+		if err != nil {
+			// we want to ignore unmarshal errors for now, proper status in importKeystore
```
Comment (rkapka) [inline validator/rpc/handlers_keymanager.go:122]: > proper status in importKeystore

I don't understand what this means

File: validator/rpc/handlers_keymanager.go [hunk 7]
```diff
@@ -19,11 +21,290 @@ import (
...
+	}
+	if len(req.Passwords) == 0 {
+		req.Passwords = make([]string, len(req.Keystores))
+	}
+
+	// req.Passwords and req.Keystores are checked for 0 length in code above.
+	if len(req.Passwords) > len(req.Keystores) {
+		req.Passwords = req.Passwords[:len(req.Keystores)]
+	}
+	if len(req.Passwords) < len(req.Keystores) {
```
Comment (rkapka) [inline validator/rpc/handlers_keymanager.go:150]: This can be an `else if`

File: validator/rpc/handlers_keymanager.go [hunk 8]
```diff
@@ -19,11 +21,290 @@ import (
...
+			}
+		}
+		http2.WriteJson(w, &DeleteKeystoresResponse{Data: sts})
+		return
+	}
+	bytePubKeys := make([][]byte, len(req.Pubkeys))
+	for i, pubkey := range req.Pubkeys {
+		key, ok := shared.ValidateHex(w, "Pubkey", pubkey, fieldparams.BLSPubkeyLength)
+		if !ok {
+			return
```
Comment (rkapka) [inline validator/rpc/handlers_keymanager.go:210]: Missing `http2.HandleError`

File: validator/accounts/cli_manager.go [hunk 9]
```diff
@@ -21,6 +23,7 @@ import (
 func NewCLIManager(opts ...Option) (*CLIManager, error) {
 	acc := &CLIManager{
 		mnemonicLanguage: derived.DefaultMnemonicLanguage,
+		inputReader:      os.Stdin,
```
Comment (rkapka) [inline validator/accounts/cli_manager.go:26]: Is this related to the scope of the PR?

File: validator/rpc/handlers_keymanager.go (hunk 8 above)
Comment (james-prysm) [reply validator/rpc/handlers_keymanager.go:210]: it does it inside the validate Hex , maybe we should change that to be more aware? and return an error instead ?

File: validator/accounts/cli_manager.go (hunk 9 above)
Comment (james-prysm) [reply validator/accounts/cli_manager.go:26]: yeah based on the reader, I'm not able to mock the input for the read to test the functionality ( it was missing unit tests before)

File: validator/node/node.go (hunk 5 above)
Comment (james-prysm) [reply validator/node/node.go:826]: I need to remove this one, instead of reregistering i am doing it immediately when the server is initiated to avoid the race I'll update this.

File: validator/accounts/accounts_delete.go (hunk 2 above)
Comment (james-prysm) [reply validator/accounts/accounts_delete.go:43]: It's to mock the input reader so we can actually test the function ( we can't test it otherwise)

File: validator/accounts/accounts_delete.go (hunk 1 above)
Comment (james-prysm) [reply validator/accounts/accounts_delete.go:28]: It's to mock the input reader so we can actually test the function ( we can't test it otherwise)

File: validator/node/node.go (hunk 4 above)
Comment (james-prysm) [reply validator/node/node.go:825]: the validator API is still under /api. i guess during the transition period there are 3 kinds of APIs

File: validator/rpc/handlers_keymanager.go (hunk 8 above)
Comment (rkapka) [reply validator/rpc/handlers_keymanager.go:210]: Ah, I didn't notice the `w` param. Might be indeed better to return the error, but it should be a different PR.

File: validator/rpc/handlers_keymanager.go (hunk 6 above)
Comment (james-prysm) [reply validator/rpc/handlers_keymanager.go:122]: updating this comment to be more clear

File: validator/rpc/handlers_keymanager.go [hunk 10]
```diff
@@ -19,11 +21,289 @@ import (
...
+	}
+
+	exportedHistory, err := s.slashingProtectionHistoryForDeletedKeys(ctx, bytePubKeys, statuses)
+	if err != nil {
+		log.WithError(err).Warn("Could not get slashing protection history for deleted keys")
+		sts := make([]*keymanager.KeyStatus, len(req.Pubkeys))
+		for i := 0; i < len(req.Pubkeys); i++ {
+			sts[i] = &keymanager.KeyStatus{
+				Status:  keymanager.StatusError,
+				Message: "Non duplicate keys that were existing were deleted, but could not export slashing protection history",
```
Comment (saolyn) [inline validator/rpc/handlers_keymanager.go:232]: This error message is incredibly confusing. This was my best guess of the meaning
```suggestion
				Message: "Could not export slashing protection history as existing non duplicate keys were deleted",
```

File: validator/rpc/handlers_keymanager.go [hunk 11]
```diff
@@ -19,11 +21,289 @@ import (
...
+// For a list of deleted keystore statuses, we check if any NOT_FOUND status actually
+// has a corresponding public key in the database. In this case, we transform the status
+// to NOT_ACTIVE, as we do have slashing protection history for it and should not mark it
+// as NOT_FOUND when returning a response to the caller.
+func (s *Server) transformDeletedKeysStatuses(
+	ctx context.Context, pubKeys [][]byte, statuses []*keymanager.KeyStatus,
+) ([]*keymanager.KeyStatus, error) {
+	pubKeysInDB, err := s.publicKeysInDB(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "Could not get public keys from DB")
```
Comment (saolyn) [inline validator/rpc/handlers_keymanager.go:260]:
```suggestion
		return nil, errors.Wrap(err, "could not get public keys from DB")
```

**Review by rauljordan — APPROVED** [review]
I went through the full PR and want to say it's an awesome piece of work. The tests are thorough, the refactor is clean, and the code reads a lot nicer with the http logic. Everything flows well in the logic for the different endpoints. I also scanned for typos in strings and the like, but could not find any

LGTM!
